### ISSUE TITLE
TRAVIS_COMMIT_RANGE includes commits unrelated to PR

### ISSUE THREAD

**Issue by jawshooah** [travis-ci/travis-ci#4596]
The commit range stored in the `TRAVIS_COMMIT_RANGE` environment variable uses the triple-dot syntax `r1...r2`, which equates to "the set of commits that are reachable from either one of `r1` or `r2` but not from both".

This does not fit the description given in the [documentation](http://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables), which claims that `TRAVIS_COMMIT_RANGE` is "the range of commits that were included in the push or pull request".

In order to be accurate to that description, the commit range should use the double-dot syntax, `r1..r2`, which equates to "commits that are reachable from `r2` excluding those that are reachable from `r1`". Otherwise, commits made to the target branch after the PR branch was created will be included in the commit range.

Refs caskroom/homebrew-cask#12903

**Comment by wking** [comment]
I'm [working around this][1] with [`${TRAVIS_COMMIT_RANGE/.../..}`][2] in my `.travis.yml`.

[1]: https://github.com/opencontainers/specs/pull/182#issuecomment-139810983
[2]: https://github.com/opencontainers/specs/commit/8831b203606aa74d2bf5274108816dde7629fa6f#diff-354f30a63fb0907d4ad57269548329e3R17

**Comment by jawshooah** [comment]
Any comment on this, maintainers?

**Comment by spuder** [comment]
+1

**Comment by LinusU** [comment]
@BanzaiMan any updates on this? ❤️ It would be nice to at least get the problem confirmed by someone working at Travis, maybe add the workaround to the documentation

**Comment by s9gf4ult** [comment]
Here the quote from `git diff --help` 

```
       git diff [--options] <commit>...<commit> [--] [<path>...]
           This form is to view the changes on the branch containing and up to the second <commit>, starting at a common ancestor of both <commit>. "git diff A...B" is equivalent to "git
           diff $(git-merge-base A B) B". You can omit any one of <commit>, which has the same effect as using HEAD instead.
```

and here from `git logrevisions` 

```
       The .. (two-dot) Range Notation
           The ^r1 r2 set operation appears so often that there is a shorthand for it. When you have two commits r1 and r2 (named according to the syntax explained in SPECIFYING
           REVISIONS above), you can ask for commits that are reachable from r2 excluding those that are reachable from r1 by ^r1 r2 and it can be written as r1..r2.

       The ... (three dot) Symmetric Difference Notation
           A similar notation r1...r2 is called symmetric difference of r1 and r2 and is defined as r1 r2 --not $(git merge-base --all r1 r2). It is the set of commits that are reachable
           from either one of r1 (left side) or r2 (right side) but not from both.
```

So `git diff` and `git log` undertstands three dots differently. In fact three dots for `git diff` is like two dots for `git log`. The common syntax which works same is `^rev1 rev2`

```
git log ^rev1 rev2
git diff ^rev1 rev2
```

will work with same commits set.

**Comment by wking** [comment]
Go away, @stale bot! ;)

**Comment by JoshDreamland** [comment]
This issue may be stale and moldy, but it remains an issue. Can someone reopen this bug?

**Comment by seldridge** [comment]
Note: there's a further underlying problem here that crops up for long-lived feature branches not rebased onto master. Travis uses a shallow clone with a default depth of 50. If the shallow clone doesn't include the originating commit where the feature branch diverged, then `$TRAVIS_COMMIT_RANGE` will always include the full repository ancestry of the feature branch going back to the first commit. Both double dot and triple dot commit ranges are predicated on having the full repository ancestry.

We hit this and are currently getting around it with the following. This checks to see if evaluation of `TRAVIS_COMMIT_RANGE` would include the first commit in the repository. If it does, then the repository is unshallowed before running any scripts:
```yaml
before_script:
  - OLDEST_SHARED=`git log --format=%H $TRAVIS_COMMIT_RANGE | tail -n1`
  - OLDEST_COMMIT=`git log --format=%H | tail -n1`
  - if [ $OLDEST_SHARED == $OLDEST_COMMIT ]; then git fetch --unshallow; fi
```

**Comment by jasonkarns** [comment]
bump

**Comment by jawshooah** [comment]
Go away, stalebot

**Comment by jasonkarns** [comment]
Way to go, stalebot 😦 

Can we get a reopen @BanzaiMan ?

**Comment by marc-h38** [comment]
"Never before was so much trouble caused by a single character to so many."

Everyone affected please thumbs this up at the top: only five right now!

**Comment by marc-h38** [comment]
tl;dr: convert `${TRAVIS_COMMIT_RANGE}` to double dots only for `git log`, not for `git diff`

> The commit range stored in the TRAVIS_COMMIT_RANGE environment variable uses the triple-dot syntax r1...r2, which equates to "the set of commits that are reachable from either one of r1 or r2 but not from both".

That's true for `git log` but as explained by @s9gf4ult , the triple dots mean something different (and very useful) for `git diff`. In fact the triple dotted `${TRAVIS_COMMIT_RANGE}` is the perfect argument for `git diff --name-only`.

**Comment by marc-h38** [comment]
> So git diff and git log undertstands three dots differently. In fact three dots for git diff is like two dots for git log.

and vice-versa!

```
          *---*---PR pull request
         /
--*---base---*----upstream

     Double dots        |  Triple dots
       upstream..PR     |        upstream...PR
------------------------+--------------------------
diff   upstream  PR "V" |  base             PR "/" 
log    ^base     PR "/" | ^base  upstream   PR "V"
```

**Comment by pprindeville** [comment]
Sounds like there's an issue with git.  Has a bug been filed against it?

**Comment by marc-h38** [comment]
Git user interface issues like this one are unfixable, imagine how many things this would break.

OK, one way to "fix" interface issues like this is to add new commands and leave the old ones as they are. For instance the new `git switch` helps you avoid the confusing and dangerous `git checkout` : https://www.infoq.com/news/2019/08/git-2-23-switch-restore/
So maybe there could be new `git helps revisions`  syntaxes that don't use dots? I don't know.

For similar backward compatibility reasons it's way too late to change TRAVIS_COMMIT_RANGE. This issue should stay closed.

On the other hand Travis could/should have additional variables, for instance Gitlab has CI_COMMIT_REF_NAME, CI_COMMIT_SHA, CI_COMMIT_BRANCH and a good few others. Feel free to file a new Travis issue if none yet.
